vowpalwabbit/array_parameters.h

```diff
@@ -0,0 +1,172 @@
+#pragma once
+#include <stdio.h>
+#include <iostream>
+#include <stdlib.h>
+#include <string.h>
+
+#ifndef _WIN32
+#include <sys/mman.h>
+#endif
+
+using namespace std;
+typedef float weight;
+
+class weight_vector;
+
+class weights_iterator_iterator
+{
+private:
+	weight* _cur;
+public:
+	weights_iterator_iterator(weight* cur)
+		: _cur(cur)
+	{ }
+	
+	weight& operator*() { return *_cur; }
+
+	weights_iterator_iterator& operator++()
+	{
+		_cur++;
+		return *this;
+	}
+
+	template<typename T>
+	weights_iterator_iterator operator+(T index) { return weights_iterator_iterator(_cur + index); }
+
+	template<typename T>
+	weights_iterator_iterator& operator+=(T index)
+	{
+		_cur += index;
+		return *this;
+	}
+
+	weights_iterator_iterator& operator=(const weights_iterator_iterator& other)
+	{
+		_cur = other._cur;
+		return *this;
+	}
+
+	bool operator==(const weights_iterator_iterator& rhs) { return _cur == rhs._cur; }
+	bool operator!=(const weights_iterator_iterator& rhs) { return _cur != rhs._cur; }
+
+};
+class weights_iterator
+{
+private:
+	weight* _current;
+	uint32_t _stride;
```

Comment:
Nitpick: Strictly speaking, this should be std::ptrdiff_t, not uint_32.
Nitpick 2: Unless there's a good reason for it, you should always prefer to use abstract types such as int, long, or std::size_t instead of the bitness specific ones.
Nitpick 3: All of your uintX_t should really be std::uintX_t (at least in header files)